NegativeArraySizeException in AbstractCodec

Occurs while decoding what I'm guessing should be a null array value for a field:

```
Caused by: java.lang.NegativeArraySizeException: -1
	at org.eclipse.milo.opcua.binaryschema.AbstractCodec.decode(AbstractCodec.java:244)
	at org.eclipse.milo.opcua.stack.core.types.OpcUaDefaultBinaryEncoding.decode(OpcUaDefaultBinaryEncoding.java:115)
	at org.eclipse.milo.opcua.stack.core.types.builtin.ExtensionObject.lambda$decode$0(ExtensionObject.java:126)
	at org.eclipse.milo.opcua.stack.core.util.Lazy.maybeCompute(Lazy.java:45)
	at org.eclipse.milo.opcua.stack.core.util.Lazy.getOrCompute(Lazy.java:35)
	at org.eclipse.milo.opcua.stack.core.types.builtin.ExtensionObject.decode(ExtensionObject.java:126)
	at org.eclipse.milo.opcua.stack.core.types.builtin.ExtensionObject.decode(ExtensionObject.java:117)
```

--- a/opc-ua-stack/bsd-parser/bsd-parser-core/src/main/java/org/eclipse/milo/opcua/binaryschema/AbstractCodec.java
+++ b/opc-ua-stack/bsd-parser/bsd-parser-core/src/main/java/org/eclipse/milo/opcua/binaryschema/AbstractCodec.java
@@ -241,7 +241,13 @@ public abstract class AbstractCodec<StructureT, MemberT> implements OpcUaBinaryD
 
                     members.put(fieldName, opcUaToMemberTypeArray(fieldName, bitAccumulation.intValue(), typeName));
                 } else {
-                    Object[] values = new Object[length];
+                    Object[] values;
+
+                    if (length >= 0) {
+                        values = new Object[length];
+                    } else {
+                        values = new Object[0];
+                    }
 
                     if (typeNamespaceIsUa && READERS.containsKey(typeName)) {
                         for (int i = 0; i < length; i++) {

--- a/opc-ua-stack/bsd-parser/bsd-parser-core/src/main/java/org/eclipse/milo/opcua/binaryschema/AbstractCodec.java
+++ b/opc-ua-stack/bsd-parser/bsd-parser-core/src/main/java/org/eclipse/milo/opcua/binaryschema/AbstractCodec.java
@@ -17,6 +17,7 @@ import java.util.Map;
 import java.util.UUID;
 import java.util.function.BiConsumer;
 import java.util.function.Function;
+import javax.annotation.Nullable;
 
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Iterators;
@@ -241,25 +242,23 @@ public abstract class AbstractCodec<StructureT, MemberT> implements OpcUaBinaryD
 
                     members.put(fieldName, opcUaToMemberTypeArray(fieldName, bitAccumulation.intValue(), typeName));
                 } else {
-                    Object[] values;
+                    Object[] values = null;
 
                     if (length >= 0) {
                         values = new Object[length];
-                    } else {
-                        values = new Object[0];
-                    }
 
-                    if (typeNamespaceIsUa && READERS.containsKey(typeName)) {
-                        for (int i = 0; i < length; i++) {
-                            Object value = READERS.get(typeName).apply(decoder);
+                        if (typeNamespaceIsUa && READERS.containsKey(typeName)) {
+                            for (int i = 0; i < length; i++) {
+                                Object value = READERS.get(typeName).apply(decoder);
 
-                            values[i] = value;
-                        }
-                    } else {
-                        for (int i = 0; i < length; i++) {
-                            Object value = context.decode(typeNamespace, typeName, decoder);
+                                values[i] = value;
+                            }
+                        } else {
+                            for (int i = 0; i < length; i++) {
+                                Object value = context.decode(typeNamespace, typeName, decoder);
 
-                            values[i] = value;
+                                values[i] = value;
+                            }
                         }
                     }
 
@@ -388,7 +387,7 @@ public abstract class AbstractCodec<StructureT, MemberT> implements OpcUaBinaryD
      * @param typeName the name of the OPC UA DataType.
      * @return a member of type {@link MemberT}.
      */
-    protected abstract MemberT opcUaToMemberTypeScalar(String name, Object value, String typeName);
+    protected abstract MemberT opcUaToMemberTypeScalar(String name, @Nullable Object value, String typeName);
 
     /**
      * Convert an OPC UA array value into a member of type {@link MemberT}.
@@ -398,7 +397,7 @@ public abstract class AbstractCodec<StructureT, MemberT> implements OpcUaBinaryD
      * @param typeName the name of the OPC UA DataType.
      * @return member of type {@link MemberT}.
      */
-    protected abstract MemberT opcUaToMemberTypeArray(String name, Object values, String typeName);
+    protected abstract MemberT opcUaToMemberTypeArray(String name, @Nullable Object values, String typeName);
 
     protected abstract Object memberTypeToOpcUaScalar(MemberT member, String typeName);
 

--- a/opc-ua-stack/bsd-parser/bsd-parser-gson/src/main/java/org/eclipse/milo/opcua/binaryschema/gson/JsonStructureCodec.java
+++ b/opc-ua-stack/bsd-parser/bsd-parser-gson/src/main/java/org/eclipse/milo/opcua/binaryschema/gson/JsonStructureCodec.java
@@ -124,21 +124,25 @@ public class JsonStructureCodec extends AbstractCodec<JsonObject, JsonElement> {
 
     @Override
     protected JsonElement opcUaToMemberTypeArray(String name, Object values, String typeName) {
-        JsonArray array = new JsonArray();
+        if (values == null) {
+            return JsonNull.INSTANCE;
+        } else {
+            JsonArray array = new JsonArray();
 
-        if (values instanceof Object[]) {
-            Object[] objects = (Object[]) values;
+            if (values instanceof Object[]) {
+                Object[] objects = (Object[]) values;
 
-            for (Object value : objects) {
-                array.add(opcUaToMemberTypeScalar(name, value, typeName));
+                for (Object value : objects) {
+                    array.add(opcUaToMemberTypeScalar(name, value, typeName));
+                }
+            } else if (values instanceof Number) {
+                // This is a bit array...
+                Number number = (Number) values;
+                return new JsonPrimitive(number);
             }
-        } else if (values instanceof Number) {
-            // This is a bit array...
-            Number number = (Number) values;
-            return new JsonPrimitive(number);
-        }
 
-        return array;
+            return array;
+        }
     }
 
     @Override

